picard/acousticbrainz/__init__.py

```diff
@@ -0,0 +1,121 @@
+import json
+import os
+import subprocess
+import sys
+
+from picard.acousticbrainz.extractor import (
+    check_extractor_version,
+    precompute_extractor_sha,
+)
+from picard.util import load_json
+from picard.webservice import ratecontrol
+
+
+ACOUSTICBRAINZ_HOST = "acousticbrainz.org"
+ACOUSTICBRAINZ_PORT = 80
+
+ratecontrol.set_minimum_delay((ACOUSTICBRAINZ_HOST, ACOUSTICBRAINZ_PORT), 1000)
+
+acousticbrainz_extractor = os.path.dirname(sys.argv[0]) + os.sep + "streaming_extractor_music" + ("" if sys.platform != "win32" else ".exe")
+acousticbrainz_extractor_sha = precompute_extractor_sha(acousticbrainz_extractor)
+acousticbrainz_extractor_version = check_extractor_version(acousticbrainz_extractor)
+
+
+def run_extractor_entry(tagger, recording_id, input_path, output_path, extractor_callback):
+    tagger.webservice.get(ACOUSTICBRAINZ_HOST,
+                          ACOUSTICBRAINZ_PORT,
+                          "/%s/low-level" % recording_id,
+                          partial(run_extractor, tagger, input_path, output_path, extractor_callback),
+                          priority=True,
+                          important=False,
+                          parse_response_type=None)
+
+
+def extractor(input_path, output_path):
+    # If an a duplicate entry doesn't exist, process the file
+    args = [acousticbrainz_extractor, input_path, output_path]
+    startupinfo = subprocess.STARTUPINFO()
+    startupinfo.dwFlags |= subprocess.STARTF_USESHOWWINDOW
+    with subprocess.Popen(args, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, startupinfo=startupinfo) as p:
+        (out, err) = p.communicate()
+        retcode = p.returncode
+        results = (output_path, retcode, out.decode("utf-8"))
+
+    # Add feature extractor sha to the output file
+    try:
+        with open(output_path, "r", encoding="utf-8") as f:
+            features = json.load(f)
+        features["metadata"]["version"]["essentia_build_sha"] = acousticbrainz_extractor_sha
+        with open(output_path, "w", encoding="utf-8") as f:
+            json.dump(features, f, indent=4)
+    except FileNotFoundError:
+        pass
+    return results
+
+
+def run_extractor(tagger, input_path, output_path, extractor_callback, response, reply, error):
+    duplicate = False
+    # Check if AcousticBrainz server answered with the json file for the recording id
+    if not error:
+        try:
+            load_json(response)
+            duplicate = True
+        except json.JSONDecodeError:
+            pass
+
+    if duplicate:
+        # If an entry for the same recording ID already exists
+        # using the same extractor version exists, skip extraction
+        results = (None, 0, "Duplicate")
+        extractor_callback(results)
+    else:
+        # Run extractor on a different thread and call the callback when done
+        future = tagger.thread_pool.submit(extractor, input_path, output_path)
+        future.add_done_callback(extractor_callback)

```

Comment:
Found out the issue: The problem is that calling the callback does not run on the main thread, and the callback then calls `submit_features`, which in turn means the webservice.post is not run on the main thread and that fails.

`tagger.thread_pool` is really not supposed to be used directly. Instead of custom handling you should use `picard,util.thread.run_task` like this:

```
run_task(partial(extractor, input_path, output_path), extractor_callback)
```

The callack then needs to accept a result and an error object, something like this:

```
def abz_extractor_callback(file, result, error):
    if error:
        # handle the error

    abz_metadata_file, result, error = result  # Now this conflicts a bit with the naming, that should be fixed
```

Actually the `extractor` function could probably be changed to throw on error, then the error would automatically end up as the error parameter in abz_extractor_callback

Reply:
It really did fix it. :D


And I found either a bug on the extractor or a borked file. I guess this is why the recordingId check was in place.

Extractor says:
musicbrainz_trackid (supposedly the recordingID) = **'235fa3a9-feb6-434c-8801-34370c63af50'**
acoustid_id = '4519f96a-b95d-4fad-bbbc-14eb6d1913fb'

Picard says:
musicbrainz_recordingid = '3e832e23-d08a-495f-82e2-8a3632947673'
musicbrainz_trackid = 'b9fc737c-db38-491f-a45f-3e987c6a8a30'
acoustid_id = '4519f96a-b95d-4fad-bbbc-14eb6d1913fb'

Reply:
> musicbrainz_trackid (supposedly the recordingID)

Yes, that's the recording ID. For historic reasons the tag is called musicbrainz_trackid, and that's what the extractor just puts into the JSON file. Picard internally calls this `musicbrainz_recordingid`.

But strange result. Could it be because of the actual recording ID not yet written to the file?

Reply:
I don't think so. I saved the file before testing, and it happened with different files. I've seen at least 3 cases while testing yesterday, but only tested with a few hundred files.

Reply:
Found the probable reason: `235fa3a9-feb6-434c-8801-34370c63af50` is an old recording ID that got merged into `3e832e23-d08a-495f-82e2-8a3632947673`. 

But that explains the issue you describe only partially.

That means when Picard would load the release it would get definitely `3e832e23-d08a-495f-82e2-8a3632947673`, even if the file had  `235fa3a9-feb6-434c-8801-34370c63af50` in the tags. But then you would see a change from old value to new value. After saving the file should have only the new value in the tags.

But maybe the file has this stored in some non-standard way? What file format is it? For ID3 tags the MusicBrainz default is to store the recording ID in a frame `UFID` with and owner identifier of "http://musicbrainz.org". But e.g. the MusicBrainz foobar2000 plugin uses `TXXX:musicbrainz_trackid ` instead.

Reply:
Uh, found another problem. The extracted features have the values inside lists... So a few of the failures were due to that...